src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs

```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
 			return _schemes;
 		}
 
-		internal static void SetReturnData(Uri data)
+		private Activity? CurrentActivity => ContextHelper.Current as Activity;
+
+		TaskCompletionSource<WebAuthenticationResult>? tcsResponse = null;
+		Uri? currentRedirectUri = null;
+
+		public bool OnResumeCallback(Intent? intent = null)
 		{
-			if (_waitingForCallbackUri == null || _completionSource == null)
+			// If we aren't waiting on a task, don't handle the url
+			if (tcsResponse?.Task?.IsCompleted ?? true)
+				return false;
+
+			var intentData = intent?.Data?.ToString();
+			if (intentData is null ||
+				currentRedirectUri is null)
 			{
-				return; // Not waiting for this
+				tcsResponse.TrySetCanceled();
+				return false;
 			}
 
-			if (data.OriginalString.StartsWith(_waitingForCallbackUri))
+			try
 			{
-				_completionSource?.TrySetResult(
-					new WebAuthenticationResult(data.OriginalString, 0, WebAuthenticationStatus.Success));
-			}
-		}
+				var intentUri = new Uri(intentData);
 
-		internal static void OnMainActivityResumed()
-		{
-			// This is called when the application activity is resumed.
-			// If it occurs before the TCS is competed, it means the user canceled the operation.
+				// Only handle schemes we expect
+				if (!CanHandleCallback(currentRedirectUri, intentUri))
+				{
+					tcsResponse.TrySetException(new InvalidOperationException($"Invalid Redirect URI, detected `{intentUri}` but expected a URI in the format of `{currentRedirectUri}`"));
```

Comment:
What should the user do if the exception happens? Can you add a "Verify this, do that, check this" in the exception text?